// tela/lib/view.js
'use strict'

const Stream    = require('./stream').Stream
const parse     = require('./parser')
const Observes  = require('../utils').Observed
const router    = require('./router')
const Template  = require('tela-template')

class View extends Observes {
  // Lifecycle
  // 1. initialize registering all streams and their callbacks
  // 2. render the view with placeholders
  // 3. as streams update, re render the view from the state cache.

  constructor(ctx) {
    super(ctx)
    this.ctx = ctx
    this.router = router
    this.template = ''
  }

  set state(val) {
    return state[this.ctx.state.path] = val
  }

  get state() {
    return state[this.ctx.state.path]
  }

  stream(str) {
    if (str) {
      var root = str.split('.')[0]
      if (this[root] instanceof Stream) {
        return this[root]
      }
    }
  }

  // gets the property from a name and optionally a context
  // default context is the current view element.
  prop(str, ctx) {
    ctx = ctx || this;
    if (this.stream(str)) {
      var spl = str.split('.');
      var add = ''; if (spl[1]) { add += '.' + spl[1].join('.') }
      return Function('self', 'return self.' + spl[0] + '.data' + add)(ctx)
    } else {
      return Function('self', 'return self.' + str)(ctx);
    }
  }

  get params() {
    return this.ctx.params
  }

  get path() {
    return this.ctx.path
  }

  disconnect() {
    for (var key in this) {
      if (this[key] instanceof Stream) {
        this[key].remove()
      }
    }
  }

  connect() {
    this._template = new Template(this.template)
    for (var key in this) {
      if (this[key] instanceof Stream) {
        this[key].listen()
      }
    }
  }

  render() {
    this._template.render()
    if (ENV.client) {
      document.getElementById('main').innerHTML = this._template.rendered
    }

    this._template.cache

    return this._template.rendered
  }

}

module.exports = View;
